python/cudf/cudf/core/udf/row_function.py

```diff
@@ -0,0 +1,178 @@
+        if i < len(sizes) - 1:
+            next_itemsize = sizes[i + 1]
+            offset = int(math.ceil(offset / next_itemsize) * next_itemsize)
+
+    # Numba requires that structures are aligned for the CUDA target
+    _is_aligned_struct = True
+    return Record(fields, offset, _is_aligned_struct)
+
+
+def row_kernel_from_template(frame, row_type, args):
+    """
+    The kernel we want to JIT compile looks something like the following,
+    which is an example for two columns that both have nulls present
+
+    def _kernel(retval, input_col_0, input_col_1, offset_0, offset_1, size):
+        i = cuda.grid(1)
+        ret_data_arr, ret_mask_arr = retval
+        if i < size:
+            rows = cuda.local.array(1, dtype=row_type)
+            row = rows[0]
+
+            d_0, m_0 = input_col_0
+            masked_0 = Masked(d_0[i], mask_get(m_0, i + offset_0))
+            d_1, m_1 = input_col_1
+            masked_1 = Masked(d_1[i], mask_get(m_1, i + offset_1))
+
+            row["a"] = masked_0
+            row["b"] = masked_1
+
+            ret = f_(row)
+
+            ret_masked = pack_return(ret)
+            ret_data_arr[i] = ret_masked.value
+            ret_mask_arr[i] = ret_masked.valid
+
+    However we do not always have two columns and columns do not always have
+    an associated mask. Ideally, we would just write one kernel and make use
+    of `*args` - and then one function would work for any number of columns,
+    currently numba does not support `*args` and treats functions it JITs as
+    if `*args` is a singular argument. Thus we are forced to write the right
+    functions dynamically at runtime and define them using `exec`.
+    """
+    # Create argument list for kernel
+    frame = supported_cols_from_frame(frame)
+
+    input_columns = ", ".join([f"input_col_{i}" for i in range(len(frame))])
+    input_offsets = ", ".join([f"offset_{i}" for i in range(len(frame))])
+    extra_args = ", ".join([f"extra_arg_{i}" for i in range(len(args))])
+
+    # Generate the initializers for each device function argument
+    initializers = []
+    row_initializers = []
+    for i, (colname, col) in enumerate(frame.items()):
+        idx = str(i)
+        if col.mask is not None:
+            template = masked_input_initializer_template
+        else:
+            template = unmasked_input_initializer_template
+
+        initializer = template.format(idx=idx)
+
+        initializers.append(initializer)
+
+        row_initializer = row_initializer_template.format(
+            idx=idx, name=colname
+        )
+        row_initializers.append(row_initializer)
+
+    # Incorporate all of the above into the kernel code template
+    d = {
+        "input_columns": input_columns,
+        "input_offsets": input_offsets,
+        "extra_args": extra_args,
+        "masked_input_initializers": "\n".join(initializers),
+        "row_initializers": "\n".join(row_initializers),
+        "numba_rectype": row_type,  # from global
```

Comment:
What does "from global" mean?

Reply:
Same question. I noticed `row_type` in the template was not formatted and that `numba_rectype` was not actually being used.

Reply:
I think I understand why `row_type` was not formatted - it's supplied in the `global_exec_context`. Still unsure why `numba_rectype`'s role.

Reply:
The key is this part of the template kernel:
```python
# Create a structured array with the desired fields
        rows = cuda.local.array(1, dtype=row_type)
```
Basically, the only way numba understands this expression is if it has an actual python object to grab onto that it knows the type of, rather than a string like the other stuff that makes its way into the template.